src/Microsoft.OData.Client/DeepInsertSaveResult.cs

```diff
@@ -0,0 +1,303 @@
+            }
+            catch (DataServiceTransportException ex)
+            {
+                InvalidOperationException exception = WebUtil.GetHttpWebResponse(ex, ref this.batchResponseMessage);
+
+                throw exception;
+            }
+        }
+
+        /// <summary>
+        /// This method processes all the changed descriptors in the entity tracker.
+        /// It loops through all the descriptors and creates relationships between the descriptors if any.
+        /// </summary>
+        /// <typeparam name="T">The type of the top-level resource we want to do a deep insert.</typeparam>
+        /// <param name="descriptors">The list of descriptors in the entity tracker.</param>
+        /// <param name="isRootObject">True if the <paramref name="objects"/> are root or top-level objects, i.e., they were passed directly by the client, otherwise false.</param>
+        /// <param name="objects">Objects of the top-level type we want to perform a deep insert on.</param>
+        /// <exception cref="InvalidOperationException"></exception>
+        internal void BuildDescriptorGraph<T>(IEnumerable<Descriptor> descriptors, bool isRootObject, params T[] objects)
+        {
+            foreach (object obj in objects)
+            {
+                EntityDescriptor topLevelDescriptor = this.RequestInfo.EntityTracker.GetEntityDescriptor(obj);
+
+                // If we do not find an EntityDescriptor in the entitytracker
+                // for any of the provided objects then we throw an exception. 
+                if (topLevelDescriptor == null)
+                {
+                    throw Error.InvalidOperation(Strings.Context_EntityNotContained);
+                }
+
+                // If it's a top-level object sent by the user.
+                if (isRootObject)
+                {
+                    // Validate that we can only have one top level entity in Deep Insert.
+                    if (bulkUpdateGraph.TopLevelDescriptors.Count > 0)
+                    {
+                        throw Error.InvalidOperation(Strings.Context_DeepInsertOneTopLevelEntity);
+                    }
+
+                    bulkUpdateGraph.AddTopLevelDescriptor(topLevelDescriptor);
+                }
+
+                foreach (Descriptor descriptor in descriptors)
+                {
+                    if (descriptor is EntityDescriptor entityDescriptor)
+                    {
+                        if (entityDescriptor.ParentEntity != null &&
+                            entityDescriptor.ParentEntity.Equals(topLevelDescriptor.Entity) &&
+                            !bulkUpdateGraph.Contains(entityDescriptor))
+                        {
+                            // We don't support delete and update in deep insert.
+                            if (entityDescriptor.State == EntityStates.Deleted || entityDescriptor.State == EntityStates.Modified)

```

Comment:
If the existing entity is unchanged it will not be part of the `ChangedDescriptors` to be considered for a deep insert.

Reply:
To clarify the question: Let's assume we have an existing entity A in the tracker. This entity is neither new, nor deleted, nor modified. It was there already. Then we create a new entity B and link it to A. And we perform a deep insert on B. What would happen?

@ElizabethOkerio If I understand what you're saying, since `A` is unchanged, it would not be in the `ChangedDescriptors`. So does that mean we'll insert `B` but leave out `A`? So the link between `B` and `A` will not be persisted on the server?

Reply:
If A exists, we don't need to insert it.
We will do a Deep update on A so as to link the new entity B

`PATCH ~/As`
```json
{
    "@odata.context":"As/$delta",
    "value":[
        {
            "@odata.type":"Namespace.B",
            "Id":1,
            "Bs@odata.delta":[
                {
                    "@odata.id":"B(1)"
                }
            ]
        }
    ]
} 
```

In OData client
```csharp
var a= context.As.Where(x => x.Id == 1).FirstOrDefault();

var b= new B{ Id = 1001 };
context.AttachTo("Bs", b);
context.AddLink(a, "Bs", b);

context.BulkUpdate(a);
```

Reply:
1) The customer should do a deep update. 
2) Bulk update supports updates, inserts, deletes

Reply:
If `BulkUpdate` supports insertions and has fewer restrictions, why should the customer ever call `DeepInsert`?